### ISSUE TITLE
Adopt or document do_sample=False for inspect-ai evals (true greedy)

### ISSUE THREAD

**Issue by niznik-dev** [niznik-dev/cruijff_kit#499]
## Summary

Surfaced from PR #496 discussion. cruijff_kit's eval pipeline currently uses `do_sample=True, temperature=1e-7` as the default near-greedy decoding mode (set in each `@task` function under `blueprints/` and `src/tools/model_organisms/`). `-M do_sample=false` is an alternative path that disables HuggingFace's sampling step entirely and uses pure argmax, producing bitwise-reproducible greedy decoding.

We've now empirically verified `do_sample=False` works correctly through the full inspect-ai HF provider pipeline on Llama-3.2-1B-Instruct with chat template enabled — both for plain generation and for the logprob/risk_scorer path. This issue is about whether to **document or adopt** that mode rather than continue to leave it implicit.

## Why this is worth deciding

| | `do_sample=True, temperature=1e-7` (current default) | `do_sample=False` (alternative) |
|---|---|---|
| Mechanism | Logits divided by 1e-7 → very peaked softmax → sample from peaked distribution | Pure argmax over raw logits |
| Determinism | Converges to argmax in practice but goes through the sampling RNG; bitwise reproducibility is hardware/seed dependent | Bitwise reproducible. No RNG involved. |
| Temperature value passed | Has to be > 0; HF errors at exactly 0 | Ignored entirely (HF prints a one-line stderr message) |
| Output | Empirically identical (100/100 + 51/51 byte-identical samples across two paired tests) | Empirically identical |
| Logprobs / `output_logits` | Raw logits returned, scorer-side softmax math identical | Same — `output_logits=True` returns the same raw logits regardless of sampling mode |

The mode that actually expresses what we want — *deterministic argmax decoding* — is `do_sample=False`. The `1e-7` workaround is a sampling dodge whose only practical advantage is that it doesn't require knowing about `-M do_sample=false`.

## Verification artifacts

- **Generation-path test**: `/scratch/gpfs/MSALGANIK/niznik/tmp/dosample_investigation/` (capitalization on 1B-Instruct, two paired SLURM jobs).
  - 100/100 per-sample completions byte-identical between baseline (do_sample=True, temp=1e-7) and nosample (do_sample=False).
  - Same match/includes accuracies, same token counts.
- **Logprob-path test**: `/scratch/gpfs/MSALGANIK/niznik/tmp/dosample_logprobs_test/` (model_organism + risk_scorer on 1B-Instruct).
  - 51/51 per-sample completions AND 51/51 per-sample risk_scores byte-identical to float precision (e.g. 0.3953566613049765).
  - Aggregate calibration metrics (mean_risk_score, ECE, risk_calibration_error, brier_score, auc_score) all identical to 3 decimals.

Both test dirs preserved for re-runs (~30s each on 1 GPU).

## Caveats / scope of verification

- Tested only on `Llama-3.2-1B-Instruct`. We did not stress-test against other model families (Llama-3.1-8B, larger Llama-3.3-70B, base models without `-Instruct`, or any non-Llama checkpoint). Plausible that quantized models or non-standard chat templates behave differently.
- Tested only with the `cruijff` conda environment as it stood on 2026-05-18 (`transformers==4.57.6`, `inspect-ai==0.3.209`). Some HF versions historically raised on the `do_sample=False` + `temperature` warning path; current env emits a stderr line only.
- The Dec 12, 2025 train/eval template-mismatch incident that gave us the `1e-7` default was a **different** bug (chat template inconsistency between training and eval, fixed by `chat_completion` adoption). That issue is unrelated to `do_sample` per se. See [the memory entry](feedback_do_sample_false_is_safe.md) on niznik's side or PR #496 conversation for the full trace.

## Possible directions

1. **Adopt `do_sample=False` as the new default** for inspect-ai evals. Either:
   - Edit each `@task` function to accept a `do_sample: bool = False` parameter that gets passed through, OR
   - Have `scaffold-inspect` always emit `-M do_sample=false` in rendered SLURM scripts.

2. **Document `do_sample=False` as a supported alternative** but keep `1e-7` as default. Add a section to the scaffold-inspect agent doc + experiment_summary.yaml template explaining the tradeoff and how to opt in.

3. **Leave the workflow as-is**. The `1e-7` default works, the empirical test shows the modes are equivalent, no behavior change needed. Just keep the verification record around.

Argument for (1): reproducibility. The current default introduces an RNG into eval pipelines that don't conceptually need one.

Argument for (2): minimal disruption. Lets advanced users opt in without forcing the change on everyone or invalidating existing logs.

Argument for (3): the empirical equivalence means the choice is mostly cosmetic; energy better spent elsewhere.

## Acceptance criteria

- A decision on one of the above three options (or a fourth — open to alternatives).
- If (1) or (2): the change scaffolded into `setup_inspect.py`, `scaffold-inspect.md`, and (for option 1) the `@task` signatures, with a regression test confirming behavior.
- If (3): close this issue with a one-line note pointing at the verification dirs for future reference.

cc @niznik-dev

**Comment by niznik-dev** [comment]
<img src="https://raw.githubusercontent.com/niznik-dev/llm-personas/main/fusions/mattie/mattie-icon.png" width="36" align="absmiddle"> 🎻 **MxC** ✳️

## Extended verification: 3B + 8B at 5,000 samples

Following up on the 1B verification in the issue body, I re-ran the same controlled comparison — single variable `-M do_sample=false`, everything else frozen — on **Llama-3.2-3B-Instruct** and **Llama-3.1-8B-Instruct**, across both eval paths, at **5,000 samples per cell**.

### Design
- **baseline** = `do_sample=True, temperature=1e-7` (current default)
- **nosample** = identical config **+ `-M do_sample=false`**
- **Generation path**: `capitalization` task / `match` scorer (5L_80P words, 5,000 unique samples)
- **Logprob path**: `model_organism` task / `risk_scorer` with `option_tokens: ["0","1"]` (`bits_majority` k15 family, 5,000 unique samples)

### Results — all four pairs byte/float-identical

| Pair | Samples | Completions | Risk scores | Aggregate metrics |
|---|---|---|---|---|
| gen_3b | 5000 | 5000/5000 | — | 4/4 identical |
| gen_8b | 5000 | 5000/5000 | — | 4/4 identical |
| logprob_3b | 5000 | 5000/5000 | 5000/5000 | 7/7 identical |
| logprob_8b | 5000 | 5000/5000 | 5000/5000 | 7/7 identical |

**20,000 paired samples, zero divergence.** The aggregate metrics checked include `mean_risk_score`, `expected_calibration_error`, `risk_calibration_error`, `brier_score`, and `auc_score` — all identical to full float precision.

### Note on 8B
A small-N pilot (51 samples, k8) had 8B-Instruct answer in chat prose rather than `0`/`1`, so the risk path came back all-NaN — equivalent across modes, but vacuous. At 5,000 samples on the k15 pool the 8B risk path is non-degenerate and the floats are identical, so that gap is closed: we now have a genuine 8B calibration comparison, and it matches.

### Scope
Verified across **1B + 3B + 8B-Instruct** now (generation + logprob/risk, at scale). Still **Llama-only** — non-Llama families remain untested, unchanged from the original caveat.

### Recommendation
Proceed with **option 1 — adopt `do_sample=False` as the eval default.** The `1e-7` is a sampling dodge that threads an RNG through eval pipelines that don't conceptually need one; `do_sample=False` is pure argmax, bitwise-reproducible, and now empirically equivalent at scale across three model sizes.

—MxC
